packages/roosterjs-editor-dom/lib/clipboard/extractClipboardItems.ts

```diff
@@ -1,131 +1,141 @@
+import {
+    ClipboardData,
+    ContentType,
+    ContentTypePrefix,
+    EdgeLinkPreview,
+    ExtractClipboardItemsOption,
+} from 'roosterjs-editor-types';
+
+// HTML header to indicate where is the HTML content started from.
+// Sample header:
+// Version:0.9
+// StartHTML:71
+// EndHTML:170
+// StartFragment:140
+// EndFragment:160
+// StartSelection:140
+// EndSelection:160
+const CLIPBOARD_HTML_HEADER_REGEX = /^Version:[0-9\.]+\s+StartHTML:\s*([0-9]+)\s+EndHTML:\s*([0-9]+)\s+/i;
+const OTHER_TEXT_TYPE = ContentTypePrefix.Text + '*';
+const EDGE_LINK_PREVIEW = 'link-preview';
+const ContentHandlers: {
+    [contentType: string]: (data: ClipboardData, value: string, type?: string) => void;
+} = {
+    [ContentType.HTML]: (data, value) =>
+        (data.rawHtml = Browser.isEdge ? workaroundForEdge(value) : value),
+    [ContentType.PlainText]: (data, value) => (data.text = value),
+    [OTHER_TEXT_TYPE]: (data, value, type?) => !!type && (data.customValues[type] = value),
+};
+
+/**
+ * Extract clipboard items to be a ClipboardData object for IE
+ * @param items The clipboard items retrieve from a DataTransfer object
+ * @param callback Callback function when data is ready
+ * @returns An object with the following properties:
+ *  types: Available types from the clipboard event
+ *  text: Plain text from the clipboard event
+ *  image: Image file from the clipboard event
+ *  html: Html string from the clipboard event. When set to null, it means there's no HTML found from the event.
+ *   When set to undefined, it means can't retrieve HTML string, there may be HTML string but direct retrieving is
+ *   not supported by browser.
+ */
+export default function extractClipboardItems(
+    items: DataTransferItem[],
+    options?: ExtractClipboardItemsOption
+): Promise<ClipboardData> {
+    const data: ClipboardData = {
+        types: [],
+        text: '',
+        image: null,
+        files: [],
+        rawHtml: null,
+        customValues: {},
+    };
+
+    const contentHandlers = { ...ContentHandlers };
+
+    if (options?.allowLinkPreview) {
+        contentHandlers[ContentTypePrefix.Text + EDGE_LINK_PREVIEW] = tryParseLinkPreview;
+    }
+
+    return Promise.all(
+        (items || []).map(item => {
+            const type = item.type;
+
+            if (type.indexOf(ContentTypePrefix.Image) == 0 && !data.image && item.kind == 'file') {
+                data.types.push(type);
+                data.image = item.getAsFile();
+                return new Promise<void>(resolve => {
+                    if (data.image) {
+                        readFile(data.image, dataUrl => {
+                            data.imageDataUri = dataUrl;
+                            resolve();
+                        });
+                    } else {
+                        resolve();
+                    }
+                });
+            } else if (item.kind == 'file') {
+                return new Promise<void>(resolve => {

```

Comment:
This is the change to add files to `ClipboardData`. I've added the fetching of files and putting it in clipboard in the promise, is there  any reason it should be before the promise?